### PR TITLE
Add LookupAs/TryLookupAs to complement Lookup/TryLookup

### LINKED ISSUE #427: Maps (propertysets) of K to Object could get an unboxing [Try]Lookup
There's a bunch of code like this:

```c++
ValueSet v = /* ... */
auto g = winrt::unbox_value<winrt::guid>(v.Lookup(L"someKey"));
auto q = winrt::unbox_value_or<winrt::hstring>(v.TryLookup(L"anotherKey"), winrt::hstring{});
```

It'd be neat if 'unbox object to type' was directly supported so I could have written:

```c++
ValueSet v = /* ... */
auto g = v.Lookup<winrt::guid>(L"someKey");
auto q = v.TryLookup<winrt::hstring>(L"anotherKey");
```

### PR DESCRIPTION
**Pull request by kennykerr**
Fixes #427 by adding unboxing helpers for maps.

### CODE REVIEW

File: strings/base_reference_produce.h [hunk 1]
```diff
@@ -377,6 +377,56 @@ WINRT_EXPORT namespace winrt
         return default_value;
     }
 
+    template <typename T>
+    auto try_unbox_value(Windows::Foundation::IInspectable const& value)
+    {
+        if constexpr (std::is_base_of_v<Windows::Foundation::IInspectable, T>)
+        {
+            return value.try_as<T>();
```
Comment (jonwis) [inline strings/base_reference_produce.h:385]: TryLookup returns std::optional<T>{} when not present, yes?  Does that .try_as<T> return nullptr if the value is empty? Or does the `if (!value)` below need to move upwards?

File: test/test/lookup_as.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        REQUIRE(m.TryLookupAs<IStringable>(L"uri").ToString() == L"http://kennykerr.ca/");
+        REQUIRE(m.TryLookupAs<AsyncStatus>(L"enum").value() == AsyncStatus::Completed);
+        REQUIRE(m.TryLookupAs<AsyncStatus>(L"underlying_type").value() == AsyncStatus::Error);
+        REQUIRE(m.TryLookupAs<guid>(L"guid").value() == guid_of<Uri>());
+        REQUIRE(m.TryLookupAs<GUID>(L"guid").value() == static_cast<GUID>(guid_of<Uri>()));
+        REQUIRE(m.TryLookupAs<hstring>(L"hstring").value() == L"Test");
+
+        // TryLookupAs edge cases
+
+        REQUIRE_FALSE(m.TryLookupAs<bool>(L"INVALID").has_value());
```
Comment (jonwis) [inline test/test/lookup_as.cpp:49]: Does "nullptr" here represent the invalid value? Or should there be a test for `m.TryLookupAs<Uri>(L"INVALID")` here?

File: strings/base_reference_produce.h (hunk 1 above)
Comment (oldnewthing) [reply strings/base_reference_produce.h:385]: On TryLookup failure, it returns `nullptr` for reference types and `std:optional{}` for value types.

File: test/test/lookup_as.cpp (hunk 2 above)
Comment (oldnewthing) [reply test/test/lookup_as.cpp:49]: `TryLookupAs<bool>` will return a `std::optional<bool>`, and we are verifying that it returns a valueless object.

In the second test, `TryLookupAs<Uri>` will return a `Uri`, and we are using the `Uri == nullptr` test to see if the `Uri` object is empty.

**Comment by mlalic** [comment]
> I preferred the proposal to add support for value types to `IInspectable.as<>` and `IInspectable.try_as<>` (treating them as unboxing), because that is useful in places other than map lookups. For example, converting XAML tags and navigation parameters.

FWIW, in addition to the Xaml examples Raymond listed, we have a new UIAutomation platform API (currently in insider preview SDK builds) that would also benefit from a uniform way of converting an `IInspectable` to a runtimeclass, COM pointer, or an unboxed value.

For example;

```c++
// today
auto elementResult = result.GetOperand(a).as<IUIAutomationElement>(); // this is a COM interface
auto textRangeResult = result.GetOperand(b).as<winrt::AutomationTextRange>(); // this is a WinRT runtimeclass
auto stringResult = winrt::unbox_value<winrt::hstring>(result.GetOperand(c));


// new hotness :)
auto elementResult = result.GetOperand(a).as<IUIAutomationElement>();
auto textRangeResult = result.GetOperand(b).as<winrt::AutomationTextRange>();
auto stringResult = result.GetOperand(c).as<winrt::hstring>();
```

The consistency looks really appealing from a consumer's perspective, IMO.

**Comment by kennykerr** [comment]
Yes, I also prefer overloading `as` and `try_as` - that was my original suggestion. The trouble is (beyond the tricky metaprogramming) that this introduces a dependency from `winrt/base.h` (where `IUnknown`/`IInspectable`/`as`/`try_as` live) to `winrt/windows.foundation.h` (where boxing support lives). And that's just a much bigger change than I'd like to do for this feature.

File: strings/base_reference_produce.h (hunk 1 above)
Comment (jonwis) [reply strings/base_reference_produce.h:385]: Thanks, try_as of nullpter is what I didn't know.

File: test/test/lookup_as.cpp (hunk 2 above)
Comment (jonwis) [reply test/test/lookup_as.cpp:49]: How about a test that validates `TryLookupAs<IInspectableDerivedThing>(L"VALUE THAT IS NOT IN THE MAP")`?
